### ISSUE TITLE
Implement sampling for APriori

### ISSUE THREAD

**Issue by ameligrana** [JaredSchwartz/RuleMiner.jl#9]
When the dataset doesn't fit in main memory it could be useful to allow to just analyze a single sample instead of all the dataset. This would require to adjust the support to `p*s` where `p` is the sample proportion and `s` is the support required. This would be approximate, but it is very simple to implement as you see (for a reference look at 6.4.1 of http://infolab.stanford.edu/~ullman/mmds/ch6.pdf)

**Comment by JaredSchwartz** [comment]
I wonder if it would be better to write this as a function which operates on the Transactions object that samples down the transactions directly. 

This would have the advantage of being able to use the the sampled transactions with other algorithms, but it would also open up the the user to error for not adjusting their support parameter.

**Comment by ameligrana** [comment]
Seems even better to me!

**Comment by ameligrana** [comment]
Ah yes sorry re-evaluating after your last sentence. Yes, that could be a problem, I don't know if there is a way to avoid that correctness trap

**Comment by ameligrana** [comment]
Maybe allowing both things could be useful, internally we will use the sampling methods on the transactions. But a method to sample transactions is also exposed publicly

**Comment by JaredSchwartz** [comment]
Yeah, I just think that you lose some of the memory usage advantages if you are sampling in the function instead of sampling the Transactions beforehand.

Obviously, the actual algorithm runs faster, but if you are worried about fitting data into memory, this doesn't really solve the issue and actually increases memory usage if you have both the original and the sampled copy in-memory.

I also feel like the correctness issue is on the user a little bit. If we adjust things behind the scenes, the outputs won't line up with what they input unless we also scale the outputs, right? Like if they put in a support parameter of 0.5 and a sampling proportion of 0.1, the output would show all rules which have a support >= 0.05, which may also not be intuitive for people who don't understand how sampling affects the results.

**Comment by ameligrana** [comment]
> Obviously, the actual algorithm runs faster, but if you are worried about fitting data into memory, this doesn't really solve the issue and actually increases memory usage if you have both the original and the sampled copy in-memory.

In my limited testing of APriori I saw a lot of memory usage in the algorithm itself, even more than the loaded dataset at some point. Is this right?

Also, I don't know how the IO stuff is working at the moment but would be an approach that samples transactions as you see them from hard memory be possible? I wrote a package for this https://github.com/JuliaDynamics/StreamSampling.jl but not sure if it's really appropriate.

P.S: Great improvements!!

**Comment by JaredSchwartz** [comment]
This is definitely interesting, though I've put off actually adding this as I stabilize the data structures I'm using.

I don't think my current apriori algorithm has any unsafe operations. It's all dynamically allocated/managed. Are you talking about the async stuff in the sync block?

As far as stream sampling, I can definitely look into this when trying to implement sampling. for data reading, I'm iterating through an MMAP array, so it's possible that we could implement some kind of streaming solution.

One other option is to get the number of items from the file by counting the delimiters, then creating a random sample out of that range, and only keeping the transactions that are in the random sample indexes

**Comment by ameligrana** [comment]
Sorry it was "In my limited testing of APriori I saw a lot of memory usage" not "In my limited testing of APriori I saw a lot of memory usafe" in that sentence  :D, I edited it now

> One other option is to get the number of items from the file by counting the delimiters, then creating a random sample out of that range, and only keeping the transactions that are in the random sample indexes

Indeed this seems the easy option, I actually would like to implement a simple type in that package to implement the sampling knowing the length, however this requires two passes of course (if the user doesn't provide the length himself at least)

**Comment by ameligrana** [comment]
Sorry and maybe I wasn't clear enough also on this, I used Spark for APriori, not your package when I saw the memory usage increase

**Comment by JaredSchwartz** [comment]
> Indeed this seems the easy option, I actually would like to implement a simple type in that package to implement the sampling knowing the length, however this requires two passes of course (if the user doesn't provide the length himself at least)

For sure. I'm already making two passes in my data loading, so it won't be that much of a lift for my package specifically

**Comment by JaredSchwartz** [comment]
> Sorry and maybe I wasn't clear enough also on this, I used Spark for APriori, not your package when I saw the memory usage increase

I'm not sure I can help you debug Spark lol. Generally though, A Priori needs to evaluate many individual slices of the data. The slice creation will likely cause you to use more memory than the original dataset. I also use slices in my implementation because using views actually results in a speed regression.

**Comment by ameligrana** [comment]
I was just saying that because maybe you use a different method which doesn't require more memory (and no worries I hate Spark I wouldn't ask anyone to do the painful debugging on Spark operations :D).

Anyway, then sampling the transactions or using a sampling fraction as an argument in the function would anyway allow some memory optimization.
